src/rvsdg/egglog_optimizer/subst.rs

```diff
@@ -42,12 +44,144 @@ fn subst_all_rule(btype: String) -> String {
     )
 }
 
+// rtjoa: We could parameterize on `above`'s type (currently, always Body) to
+// support substituting under a Function as well
+fn subst_beneath_rules() -> Vec<String> {
+    let mut res = vec!["
+        (relation can-subst-Expr-beneath (Body Expr Expr))
+        (relation can-subst-Operand-beneath (Body Operand Operand))
+        (relation can-subst-Body-beneath (Body Body Body))
+        (relation can-subst-VecVecOperand-beneath (Body VecVecOperand VecVecOperand))
+        (relation can-subst-VecOperand-beneath (Body VecOperand VecOperand))
```

Comment:
Is the "can" part because it might not do the substitution? It might just find some version of the program that doesn't use the "from" at all?

Reply:
I'm just phrasing like a predicate, since the way to think about this is "propagating knowledge about where it is safe to subst" rather than something functional